Revamp Google Cal warning for Meet, Amie, and Vimcal

joeauyeung opened the pull request:
## What does this PR do?

<!-- Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change. -->

This PR redesigns the "Requires Google Calendar" warning for Meet, Amie, and Vimcal

Fixes #7254, fixes #7029

## Current
<img width="495" alt="image" src="https://user-images.githubusercontent.com/65426560/220999784-1ecf3ef7-60b4-4d9d-bb66-d763e971508d.png">


## With no GCal connected
<img width="495" alt="image" src="https://user-images.githubusercontent.com/65426560/220999633-261af572-ade6-49e2-854c-0ad97b3ca7de.png">


## With GCal connected
<img width="495" alt="image" src="https://user-images.githubusercontent.com/65426560/220999560-9e606a86-45af-45a8-b942-2ce9a9630ab0.png">


Fixes # (issue)

<!-- Please provide a loom video for visual changes to speed up reviews
 Loom Video: https://www.loom.com/
-->

**Environment**: Staging(main branch) / Production

## Type of change

<!-- Please delete options that are not relevant. -->

- [x] Chore (refactoring code, technical debt, workflow improvements)

## How should this be tested?

<!-- Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration -->

- Ensure that the "requires GCal" warning appears on the Meet, Vimcal, and Amie app pages
    - Make sure that the install button is disabled
- Add a GCal credential
- You should be able to install the previously blocked apps

## Checklist

<!-- Please remove all the irrelevant bullets to your PR -->

- I haven't read the [contributing guide](https://github.com/calcom/cal.com/blob/main/CONTRIBUTING.md)
- My code doesn't follow the style guidelines of this project
- I haven't performed a self-review of my own code and corrected any misspellings
- I haven't commented my code, particularly in hard-to-understand areas
- I haven't checked if my PR needs changes to the documentation
- I haven't checked if my changes generate no new warnings
- I haven't added tests that prove my fix is effective or that my feature works
- I haven't checked if new and existing unit tests pass locally with my changes

joeauyeung committed:
Create requires Google Cal component

joeauyeung committed:
Create installed GCal message

joeauyeung committed:
Move requires GCal component to App

joeauyeung commented on packages/app-store/googlevideo/components/InstallAppButton.tsx:
```diff
@@ -1,64 +1,20 @@
 
 import useAddAppMutation from "../../_utils/useAddAppMutation";
 
 export default function InstallAppButton(props: InstallAppButtonProps) {
```
Revert this back to a normal install button

hariombalhara replied:
We can delete this file itself. If InstallAppButton.tsx file doesn't exist, default InstallAppButton component is used located at https://github.com/calcom/cal.com/blob/670520c282032816555b9cb309fe2d566ae7b754/packages/app-store/components.tsx#L30

joeauyeung commented on apps/web/components/apps/App.tsx:
```diff
@@ -76,6 +81,14 @@ const Component = ({
     }
   );
 
+  const requiresGCal = appsThatRequiresGCal.some((app) => slug === app);
```
See if the app requires GCall

joeauyeung commented on apps/web/components/apps/App.tsx:
```diff
@@ -76,6 +81,14 @@ const Component = ({
   );
 
+  const requiresGCal = appsThatRequiresGCal.some((app) => slug === app);
+  const gCalInstalled = trpc.viewer.appsRouter.checkForGCal.useQuery(undefined, {
```
The return type is a boolean or undefined

joeauyeung commented on packages/app-store/googlecalendar/components/ExistingGoogleCal.tsx:
```diff
@@ -0,0 +1,54 @@
+import Link from "next/link";
```
Decided to move this component out of the Meet package and into GCal since we are using it in multiple apps

hariombalhara replied:
Considering we move the requiresApp to config.json, we should make this component generic.
If I just change the text it can be reused for Typeform app as well which requires Routing Forms
<img width="713" alt="Screenshot 2023-02-24 at 1 37 12 PM" src="https://user-images.githubusercontent.com/1780212/221125831-6805e024-6f9e-44a8-a8ce-cd202924ab51.png">

PeerRich replied:
yes i think the more generic the better. we will have the first app soon that requires Cal Video, too!

joeauyeung committed:
Clean up

PeerRich commented on apps/web/components/apps/App.tsx:
```diff
@@ -24,6 +25,9 @@ import {
 } from "@calcom/ui/components/icon";
 
+/* These app slugs all require Google Cal to be installed */
+const appsThatRequiresGCal = ["google-meet", "amie", "vimcal"];
```
can we somehow move this into app specific configs?

ideally i dont want to hard code app things into core components 

@hariombalhara any idea?

joeauyeung replied:
Perhaps adding a property to the app's metadata?

PeerRich replied:
yep, easy


we may in the future have other apps that require other apps

hariombalhara replied:
We have a similar situation right now which is Typeform requiring Routing Forms, but in that case we didn't need such a config as Typeform App's UI itself exists in Routing Form App's UI.

@joeauyeung  Yeah, that should be the right place.

hariombalhara commented:
I can install Google Meet without installing Google Calendar from Apps Listing page
![image](https://user-images.githubusercontent.com/1780212/221125432-c324f061-33cf-4f56-a671-5d996d912b94.png)

hariombalhara reviewed (changes requested):
Left some comments.

joeauyeung committed:
Abstract prerequisite component

joeauyeung committed:
Add requires message on app card

joeauyeung committed:
Merge branch 'main' into revamp-google-cal-warning

joeauyeung committed:
Refactor to dependency

joeauyeung commented on apps/web/components/apps/App.tsx:
```diff
@@ -203,6 +220,22 @@ const Component = ({
+        {dependency &&
+          (!dependencyInstalled.isLoading ? (
+            <div className="mt-6">
+              <AppDependencyComponent
+                appName={name}
+                dependency={dependency}
+                dependencyName={dependencyInstalled.data?.dependencyName}
+                dependencyInstalled={dependencyInstalled.data?.dependencyInstalled}
+                isLoading={dependencyInstalled.isLoading}
+              />
```
Abstracted the dependency component

joeauyeung commented on packages/app-store/_appRegistry.ts:
```diff
@@ -75,13 +76,24 @@ export async function getAppRegistryWithCredentials(userId: number) {
   const apps = [] as (App & {
     credentials: Credential[];
     isDefault?: boolean;
+    dependencyData?: {
+      name?: string;
+      installed?: boolean;
+    };
```
When getting all apps for the app store we need to check if the app has a dependency. If so, prevent installing and direct the user to install the dependency

joeauyeung commented on packages/app-store/googlevideo/components/ExistingGoogleCal.tsx:
```diff
@@ -1,44 +0,0 @@
-import { Trans } from "next-i18next";
```
Abstracted to a reusable component

joeauyeung commented on packages/ui/components/apps/AppCard.tsx:
```diff
@@ -75,6 +81,14 @@ export function AppCard({ app, credentials, searchText }: AppCardProps) {
         {app.description}
       </p>
 
+      {app.dependencyData && !app.dependencyData?.installed && (
+        <Link
+          className="mt-5 rounded-md bg-blue-100 p-0.5 text-sm font-semibold text-blue-900 underline"
+          href={`${CAL_URL}/apps/${app.dependency}`}>
+          {t("requires_app", { dependencyName: app.dependencyData.name })}
+        </Link>
+      )}
```
Display this on the app card if the dependency is not installed. @Jaibles thoughts?

![image](https://user-images.githubusercontent.com/65426560/221754161-411572d7-97a3-4931-9610-5d0b4c34c250.png)

joeauyeung replied:
If Typeform requires Routing Form, should we move Typeform to EE? I have a feeling it might cause confusion if someone wants to install Typeform.

hariombalhara replied:
Moving to ee is not required now. In fact we have to move routing forms out of ee. 
License component tells the user if it's enterprise or not.

joeauyeung reviewed (commented):
@hariombalhara abstracted the app dependency component.

![image](https://user-images.githubusercontent.com/65426560/221753796-b60170b6-9af6-4fa7-8160-4646c29c3ef4.png)
![image](https://user-images.githubusercontent.com/65426560/221754129-e75eba6e-ce9f-4992-9a10-9d709a62eb1e.png)

joeauyeung committed:
Clean up

Jaibles commented:
@joeauyeung ui changes above look good. 

For this below though, we need to move this banner out of here and handle it after they click install somehow.  It doesn't fit well and looks off. 

![image](https://user-images.githubusercontent.com/4536123/221792773-482e696d-a7c6-4165-8756-f3d4b7d14e3f.jpeg)

joeauyeung committed:
Change typeform dep & remove app card dep component

joeauyeung commented:
> @joeauyeung ui changes above look good.
> 
> For this below though, we need to move this banner out of here and handle it after they click install somehow. It doesn't fit well and looks off.
> 
> ![image](https://user-images.githubusercontent.com/4536123/221792773-482e696d-a7c6-4165-8756-f3d4b7d14e3f.jpeg)


In this case, would it be better to disable the "Install" button and have the user navigate to the app page or remove the install button all together and have just the "Details" button.

![image](https://user-images.githubusercontent.com/65426560/222036591-44cb15c0-2b4f-4019-bd7e-6a5e9804eb19.png)

![image](https://user-images.githubusercontent.com/65426560/222036672-de6044f7-1a55-4782-9cf3-9c80a6b46c2c.png)

Jaibles commented:
@joeauyeung  I like the disabled install button. They can then click details, and that is where we'll show them the banner with the full context as we're currently doing.

PeerRich commented:
hold on: i honestly dont think that routing forms should require typeform. its optional in my opinion.

hariombalhara commented:
@PeerRich already left a comment here

joeauyeung committed:
Clean up

hariombalhara commented on packages/app-store/typeform/components/InstallAppButton.tsx:
```diff
@@ -0,0 +1,19 @@
+import type { InstallAppButtonProps } from "@calcom/app-store/types";
```
Why do we need to add this file. InstallAppButton.tsx isn;t a required file and the install app button can work without it.
My aim is to remove all existing such files. Any new app created doesn't have this file.

joeauyeung committed:
Change dependency to dependencies

joeauyeung committed:
Pass disableInstall to default install button for AppCard

joeauyeung committed:
Refactor app page to dependencies

joeauyeung committed:
Merge branch 'main' into revamp-google-cal-warning

joeauyeung committed:
Type fix

joeauyeung committed:
More type fixes

joeauyeung commented:
@hariombalhara I removed the `InstallAppButton` from Typeform, passed the disabled prop to the install button and refactored it from one dependency to multiple if needed.

PeerRich committed:
Merge branch 'main' into revamp-google-cal-warning

hariombalhara commented on apps/web/components/apps/App.tsx:
```diff
@@ -24,6 +24,9 @@ import {
 } from "@calcom/ui/components/icon";
 
+/* These app slugs all require Google Cal to be installed */
+const appsThatRequiresGCal = ["google-meet", "amie", "vimcal"];
```
```suggestion
```

hariombalhara commented on apps/web/components/apps/App.tsx:
```diff
@@ -76,6 +80,15 @@ const Component = ({
   );
 
+  const dependencyData = trpc.viewer.appsRouter.queryForDependencies.useQuery(dependencies, {
+    enabled: !!dependencies,
```
Can dependencies by empty. We can avoid that as well

hariombalhara commented on packages/app-store/_appRegistry.ts:
```diff
@@ -93,7 +108,24 @@ export async function getAppRegistryWithCredentials(userId: number) {
+      ...(app.dependencies && { dependencyData }),
     });
+
+    if (app.slug === "typeform") {
+      console.log({
+        rating: rating || 0,
+        reviews: reviews || 0,
+        trending: trending || true,
+        verified: verified || true,
+        ...remainingAppProps,
+        categories: dbapp.categories,
+        credentials: dbapp.credentials,
+        installed: true,
+        isDefault: usersDefaultApp === dbapp.slug,
+        ...(app.dependencies && { dependencyData }),
+      });
+    }
```
```suggestion
```

hariombalhara committed:
Merge remote-tracking branch 'origin/main' into revamp-google-cal-warning

hariombalhara committed:
Merge remote-tracking branch 'origin/revamp-google-cal-warning' into revamp-google-cal-warning

hariombalhara reviewed (commented):
Left some more comments but it looks good to merge. Testing locally now.

hariombalhara committed:
Update apps/web/components/apps/App.tsx

hariombalhara committed:
Apply suggestions from code review

hariombalhara commented:
Clicking on connect button takes me to 404 page https://www.loom.com/share/89a35609b3f74c2885b76ad587231825.

hariombalhara commented:
nit: There should be a tooltip here to tell why the button is disabled.
<img width="394" alt="Screenshot 2023-03-07 at 6 35 13 PM" src="https://user-images.githubusercontent.com/1780212/223430418-39b76c33-bb56-44f2-a008-5aabdd9e65cc.png">

PeerRich commented:
yeah follow up PR

hariombalhara commented:
<img width="403" alt="Screenshot 2023-03-07 at 6 37 15 PM" src="https://user-images.githubusercontent.com/1780212/223430885-8ee50d13-3bab-41a7-9350-68fad16f5b2a.png">
Installed button is enabled for Google Meet even when I don't have Google Calendar installed. This works correctly for Amie and Vimcal which share the same dependency.

PeerRich commented:
or maybe just hide the button

PeerRich committed:
Merge branch 'main' into revamp-google-cal-warning

joeauyeung committed:
Remove Google Meet button

joeauyeung commented:
> <img alt="Screenshot 2023-03-07 at 6 37 15 PM" width="403" src="https://user-images.githubusercontent.com/1780212/223430885-8ee50d13-3bab-41a7-9350-68fad16f5b2a.png">
> Install button is enabled for Google Meet even when I don't have Google Calendar installed. This works correctly for Amie and Vimcal which share the same dependency.

@hariombalhara removed the install app button for Google Meet just to use the default app button.

As we talked about, in a follow-up PR I can remove other `AppInstallButtons` from other app packages.

joeauyeung committed:
Merge branch 'main' into revamp-google-cal-warning

joeauyeung commented:
> or maybe just hide the button

@PeerRich I asked @Jaibles about it and he liked disabling the install button
https://github.com/calcom/cal.com/pull/7308#issuecomment-1449975866

joeauyeung committed:
Fix broken slug

hariombalhara commented:
After installation, Install button get's disabled which doesn't match with other app's behaviour where it get's removed.
See, Amie vs Around
<img width="1151" alt="Screenshot 2023-03-09 at 11 46 56 AM" src="https://user-images.githubusercontent.com/1780212/223936617-63c02de0-9f9a-45f9-acd4-d558c1e1d7f3.png">

hariombalhara committed:
Merge remote-tracking branch 'origin/main' into revamp-google-cal-warning

hariombalhara committed:
Fix TS error

hariombalhara committed:
Fix Lint errors
